src/zmqports.cpp

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) 2009-2010 Satoshi Nakamoto
+// Copyright (c) 2014 The Bitcoin developers
+// Distributed under the MIT/X11 software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "zmqports.h"
+
+#if ENABLE_ZMQ
+#include <zmq.h>
+#endif
+
+#include "primitives/block.h"
+#include "primitives/transaction.h"
+#include "version.h"
+#include "streams.h"
+#include "util.h"
+#include "netbase.h"
+#include "ui_interface.h"
+#include "json/json_spirit_value.h"
+
+using namespace json_spirit;
+using namespace std;
+
+// Global state
+bool fZMQPub = false;
+
+#if ENABLE_ZMQ
+
+// ZMQ related file scope variables
+static void *zmqContext;
+static void *zmqPubSocket;
+
+// Internal utility functions
+static void ZMQPublishBlock(const CBlock &blk);
+static void ZMQPublishTransaction(const CTransaction &tx);
+
+static void zmqError(const char *str)
+{
+  LogPrint("ZMQ error: %s, errno=%s\n", str, zmq_strerror(errno));
+}
+
+// Called at startup to conditionally set up ZMQ socket(s)
+void ZMQInitialize(const std::string &endp)
+{
+  zmqContext = zmq_init(1);
+  if (!zmqContext) {
+    zmqError("Unable to initialize ZMQ context");
+    return;
+  }
+
+  zmqPubSocket = zmq_socket(zmqContext, ZMQ_PUB);
+  if (!zmqPubSocket) {
+    zmqError("Unable to open ZMQ pub socket");
+    return;
+  }
+
+  int rc = zmq_bind(zmqPubSocket, endp.c_str());
+  if (rc != 0) {
+    zmqError("Unable to bind ZMQ socket");
+    zmq_close(zmqPubSocket);
+    zmqPubSocket = 0;
+    return;
+  }
+
+  uiInterface.NotifyAcceptBlock.connect(ZMQPublishBlock);
+  uiInterface.NotifyRelayTx.connect(ZMQPublishTransaction);
```

Comment:
Does it make sense to use the uiInterface `CClientUIInterface` as signaling layer for zmq? Or would it be better to introduce a custom signaling class/struct?